Congestion Control section (#66)
Haivision/srt-rfc: draft-sharabayko-mops-srt.md

congestion control (LiveCC) and file transfer congestion control
(FileCC).

## SRT Packet Pacing and Live Congestion Control (LiveCC) {#liveCC}

To ensure smooth video playback of a receiving peer, SRT in live mode
({{transmission-mode-live}}) must control the sender's buffer level to
prevent overfill and depletion. Output pace control (OPC) module is
designed to send packets as fast as they are submitted by the
application while maintaining relatively stable buffer level.
While this looks like a simple problem, the sender's constituent
of the Automatic Repeat Request (ARQ) system between
input and output of the SRT sender adds some complexity.

SRT needs a certain amount of bandwidth overhead in order to have
space to insert retransmitted packets without affecting the SRT
sender's main stream output rate too much so that packets can be
transmitted normally.

In live mode, the timing of sending original packets is defined by
the live input (e.g., encoder or streaming application).
The goal of OPC module is to recreate the sender's input on the decoder
(or receiving application) output, such that the act of retransmitting
packets is transparent, as if they never had to be resent in

> the first place.

Author: I will delete this paragraph then.

It is done by adjusting the maximum allowed bandwidth MAX_BW

> ({{maxbw}}) which limits the bandwidth usage by SRT and is used

Author: I think my first version is good - it limits the bandwidth usage of SRT by means of introducing maximum allowed bandwidth MAX_BW. Otherwise, "maximum allowed" is written twice in one sentence and it complicates it. I will keep the original version.

by live congestion control (LiveCC) module to calculate the minimum
inter-sending time of consecutive packets PKT_SND_PERIOD.
In principle, the space between packets determines where
retransmissions can be inserted, and the overhead represents the
available margin. There is an empiric calculation that defines the
interval (in microseconds) between two packets to give a certain
bitrate. It's a function of the average packet payload (which
includes video, audio, etc.) and configured maximum bandwidth MAX_BW.
See section {{default-liveCC}} for details.

> 
> Also it is allowed to drop those packets that can not be delivered
> in time ({{too-late-packet-drop}}).

Reviewer: A hanging paragraph that loses reference with the context.
```suggestion

It is allowed in live mode to drop those packets that can not be delivered
in time ({{too-late-packet-drop}}).
```

To sum up, a combination of output pace control (OPC) and live
congestion control (LiveCC) modules controls SRT packets pacing based
on the input rate and an overhead for packets retransmission in order
to avoid congestion during fluctuations of the source bitrate.

In addition to these mechanisms, in highly variable networks live mode
fairness can be accomplished outside SRT by controlling the bitrate
of the encoder at the input of the SRT sender. SRT provides a variaty
of network related statistics such as RTT estimation, the level of

> packet loss, the number of dropped by SRT packets, etc., which can be

Reviewer: ```suggestion
packet loss, the number of packets dropped by SRT, etc., which can be
```

used for making decisions and adjusting the bitrate in real time.

### Configuring Maximum Bandwidth {#maxbw}

There are several ways of configuring maximum bandwidth MAX_BW:

1. By means of setting the value of maximum bandwidth MAX_BW
explicitly (MAXBW mode).

The recommended default value is 1 Gbps. Note that the default value
is set only for live transmission mode.

Note that this static setting is not well-suited to a variable input,
like when you change the bitrate on an encoder. Each time the input
bitrate is configured on the encoder, MAX_BW must also be reconfigured.

In this mode, SRT does not follow the fluctuations in the encoder input
making it not suitable for variable bitrate (VBR) encoding use cases.

2. By means of setting the SRT sender's input rate INPUT_BW and the

> overhead OVERHEAD (INPUTBW mode).

Reviewer: ### TODO:

- [ ] Fix list enumeration in TXT output (all numbers produce "1").

3. Measuring SRT sender's input rate internally and setting the
overhead OVERHEAD (ESTINPUTBW mode).

In this mode, SRT adjusts the value of maximum bandwidth each time it
gets the updated estimation of input rate EST_INPUT_BW:

~~~
MAX_BW = EST_INPUT_BW * (1 + OVERHEAD /100)
~~~

4. A hybrid mode which takes the configured input rate INPUT_BW,
compares it against estimated value EST_INPUT_BW, and uses the
maximum of two (HYBRIDBW mode):

~~~
MAX_BW = max(INPUT_BW, EST_INPUT_BW) * (1 + OVERHEAD /100)
~~~

Note that the units of MAX_BW, INPUT_BW, and EST_INPUT_BW are bytes per
second. OVERHEAD is defined in %.

Both ESTINPUTBW and HYBRIDBW modes are recommended for setting the
maximum bandwidth as it (MAX_BW) follows the fluctuations in
SRT sender's input rate. However, there are certain considerations
that should be taken into account.

In INPUTBW mode, SRT takes as an input the rate that had been
configured as the expected output rate of an encoder (in terms
of bitrate for the packets including audio and overhead). But it is
normal for an encoder to occasionally overshoot. At low bitrate,
sometimes an encoder can be too optimistic and will output more bits

> than expected. Under these conditions, SRT packets would not go out
> fast enough because SRT ends up being wrongly configured.

Reviewer: ```suggestion
than expected. Under these conditions, SRT packets would not go out
fast enough due to too low value of the configured bandwidth limitation.
```

where MAX_BW is the configured maximum bandwidth which limits the
bandwidth usage by SRT, in bytes per second; PKT_SND_PERIOD is
measured in microseconds.

On a retransmission timeout (RTO) event, follow the same steps as
described in "On an acknowledgement (ACK) packet reception" part.

RTO is the time that an acknowledgement is expected after a data packet
is sent out. If there is no ACK after this amount of time elapsed, a
timeout event is triggered. Since SRT only acknowledges every SYN time
(see section {{packet-acks}}), the value of retransmission timeout is
defined as follows:

~~~
RTO = RTT + 4 * RTTVar + 2 * SYN
~~~

where RTT is the round-trip time estimation, in microseconds, and
RTTVar is the variance of RTT estimation, in microseconds, reported
by the receiver and smoothed at the sender side (see sections
{{ctrl-pkt-ack}}, {{rtt}}).

Continuous timeout should increase the RTO value. In SRT, a counter
(RexmitCount) is used to track the number of continuous timeout:

~~~
RTO = RexmitCount * (RTT + 4 * RTTVar + 2 * SYN) + SYN
~~~

On the receiver side, on a loss report sending, the interval of sending
periodic NAK reports ({{packet-naks}}) is updated as follows:

~~~

> NAKInterval = (RTT + 4 * RTTVar) / 2

Author: Yes, sure, I will make a note about it.

The algorithm controls sending rate by tuning packet sending period,
i.e. how often to send out the packets. The sending rate is increased
upon an acknowledgement (ACK) reception and decreased in case of
receiving a loss report (negative acknowledgement, or NAK).
Only full ACKs (not light) trigger the increase of sending rate.

The congestion control includes two phases. It starts with a slow
start phase where it probes the network to determine available
bandwidth and the target sending rate for the desired operation
mode: congestion avoidance. In this phase, if there is no congestion
detected via loss reports, the sending rate is gradually increased.
On the other hand, if a network congestion is detected, the algorithm
slows down and decreases the sending rate to reduce following packet
loss. The slow start phase runs exactly once at the beginning of a
connection and stops when a packet loss occurs, congestion window
size reaches its maximum value, or on a timeout event.

The detailed algorithm behaviour at both phases is described
in sections {{default-fileCC-slow-start}} and
{{default-fileCC-congestion-avoidance}}, respectively.

#### Slow Start {#default-fileCC-slow-start}

> During the slow start phase, the packet sending period PKT_SND_PERIOD
> is kept 1 microsecond in order to send packets as fast as possible.

Reviewer: ```suggestion
During the slow start phase, the packet sending period PKT_SND_PERIOD
is kept at 1 microsecond in order to send packets as fast as possible, but not with infinite rate.
```

The initial value of congestion window size CWND_SIZE is set to 16
packets. CWND_SIZE has an upper threshold, the maximum allowed
congestion window size MAX_CWND_SIZE, which is used for safety purposes
(even if there is no packet loss, the slow start phase has to stop
at a certain point). The threshold can be set to the maximum of
the receiver buffer size (12 MB).

On an acknowledgement (ACK) packet reception:

Step 1. If the time passed since the last time LastRCTime the sending rate
was either increased, or kept, is less than RC_INTERVAL:

a. Keep the sending rate at the same level;

b. Stop.

>     <CODE BEGINS>
>     if (to_microseconds(currTime - LastRCTime) < RC_INTERVAL)
>     {
>         Keep the sending rate at the same level;
>         Stop;
>     }
>     <CODE ENDS>

Reviewer: ```suggestion
    <CODE BEGINS>
    if (currTime - LastRCTime < RC_INTERVAL)
    {
        Keep the sending rate at the same level;
        Stop;
    }
    <CODE ENDS>
```

reported by the receiver and smoothed at the sender side (see sections
{{ctrl-pkt-ack}}, {{rtt}});
- RC_INTERVAL is the fixed rate control interval, in microseconds.
RC_INTERVAL of SRT is SYN, or synchronization time interval, which
is 0.01 second. An ACK in SRT is sent every fixed time interval.
The maximum and default ACK time interval is SYN.
See section {{packet-acks}} for details.

On a loss report (NAK) packet reception:
- Slow start phase ends;
- Set the packet sending period PKT_SND_PERIOD as described in Step 5,
"On an acknowledgement (ACK) packet reception" part.

On a retransmission timeout (RTO) event:
- Slow start phase ends;
- Set the packet sending period PKT_SND_PERIOD as described in Step 5,
"On an acknowledgement (ACK) packet reception" part.

#### Congestion Avoidance {#default-fileCC-congestion-avoidance}

Once the slow start phase ends, the algorithm enters the congestion
avoidance phase and behaves as described below.

On an acknowledgement (ACK) packet reception:

Step 1. If the time passed since the last time LastRCTime the sending rate
was either increased, or kept, is less than RC_INTERVAL:

a. Keep the sending rate at the same level;

b. Stop.

>     <CODE BEGINS>
>     if (to_microseconds(currTime - LastRCTime) < RC_INTERVAL)
>     {
>         Keep the sending rate at the same level;
>         Stop;
>     }
>     <CODE ENDS>

Reviewer: `to_microseconds` is implementation-specific.
```suggestion
    <CODE BEGINS>
    if (currTime - LastRCTime < RC_INTERVAL)
    {
        Keep the sending rate at the same level;
        Stop;
    }
    <CODE ENDS>
```

When an SRT socket receives data, it updates its local RTT and RTTVar,
which can be used for its own sender as well.

# SRT Packet Pacing and Congestion Control

> SRT provides certain mechanisms for the sender to get a feedback
> from the receiving side through acknowledgement (ACK) packets
> ({{ctrl-pkt-ack}}). Every 10 milliseconds the sender receives
> the latest values of RTT, RTT variance, available buffer size,
> receiving rate and estimated link capacity. Upon reception of
> an NAK packet ({{ctrl-pkt-nak}}) the sender detects packet loss
> during the transmission. These mechanisms provide a solid background
> for various congestion control algorithms integration in the
> SRT protocol.

Reviewer 2: ```suggestion
SRT provides certain mechanisms for exchanging feedback on the state of 
packet transmission between sender  and receiver. Every 10 milliseconds 
the receiving side sends acknowledgement (ACK) packets ({{ctrl-pkt-ack}}) 
to the sender that include the latest values of RTT, RTT variance, available 
buffer size, receiving rate, and estimated link capacity. Similarly, NAK packets 
({{ctrl-pkt-nak}}) from the receiver inform the sender of any packet loss during 
the transmission, triggering an appropriate response. These mechanisms provide 
a solid background for the integration of various congestion control algorithms 
in the SRT protocol.
```

> Given that SRT can operate in live and file transfer modes, there are
> two groups of congestion control algorithms defined in SRT: live
> congestion control (LiveCC) and file transfer congestion control
> (FileCC).

Reviewer 2: ```suggestion
There are two groups of congestion control algorithms defined in SRT, 
corresponding to the two modes of operation: live congestion control (LiveCC), 
and file transfer congestion control (FileCC).
```

Author: As SRT is designed both for live streaming and file/message transmission {{link}}, there are two groups of congestion control algorithms defined in SRT: live congestion control (LiveCC), and file transfer congestion control (FileCC).

Author: we have two modes in SRT: message and buffer, it's not live and file transfer as I've figured out. Live streaming is possible with using message mode of SRT, file transfer - both message and buffer. We will be reworking [Data Transmission Modes](https://tools.ietf.org/html/draft-sharabayko-mops-srt-01#page-27) section a bit.

Reviewer 2: Yes, your change makes it better.

## SRT Packet Pacing and Live Congestion Control (LiveCC) {#liveCC}

> To ensure smooth video playback of a receiving peer, SRT in live mode
> ({{transmission-mode-live}}) must control the sender's buffer level to
> prevent overfill and depletion. Pacing control module is
> designed to send packets as fast as they are submitted by the
> application while maintaining relatively stable buffer level.
> While this looks like a simple problem, the sender's constituent
> of the Automatic Repeat Request (ARQ) system between
> input and output of the SRT sender adds some complexity.

Reviewer 2: ```suggestion
To ensure smooth video playback on a receiving peer, SRT in live mode
({{transmission-mode-live}}) must control the sender's buffer level to
prevent overfill and depletion. The pacing control module is
designed to send packets as fast as they are submitted by a
video application while maintaining a relatively stable buffer level.
While this looks like a simple problem, the details of the Automatic 
Repeat Request (ARQ) behaviour between input and output of the 
SRT sender add some complexity.
```

> SRT needs a certain amount of bandwidth overhead in order to have
> space to insert retransmitted packets without affecting the SRT
> sender's main stream output rate too much so that packets can be
> transmitted normally.

Reviewer 2: ```suggestion
SRT needs a certain amount of bandwidth overhead in order to have
space for the sender to insert packets for retransmission with minimum 
impact on the output rate of the main packet transmission.
```

> It is done by adjusting the maximum allowed bandwidth MAX_BW
> ({{maxbw}}) which limits the bandwidth usage by SRT and is used

Reviewer 2: ```suggestion
This balance is achieved by adjusting the maximum allowed bandwidth MAX_BW
({{maxbw}}) which limits the bandwidth usage by SRT. The MAX_BW value is used
```

> by live congestion control (LiveCC) module to calculate the minimum
> inter-sending time of consecutive packets PKT_SND_PERIOD.
> In principle, the space between packets determines where
> retransmissions can be inserted, and the overhead represents the
> available margin. There is an empiric calculation that defines the
> interval, in microseconds, between two packets to give a certain
> bitrate. It's a function of the average packet payload (which
> includes video, audio, etc.) and configured maximum bandwidth MAX_BW.
> See {{default-liveCC}} for details.

Reviewer 2: ```suggestion
by the live congestion control (LiveCC) module to calculate the minimum
interval between consecutive sent packets PKT_SND_PERIOD.
In principle, the space between packets determines where
retransmissions can be inserted, and the overhead represents the
available margin. There is an empiric calculation that defines the
interval, in microseconds, between two packets to give a certain
bitrate. It's a function of the average packet payload (which
includes video, audio, etc.) and the configured maximum bandwidth MAX_BW.
See {{default-liveCC}} for details.
```

> In live mode, it is allowed to drop those packets that can not be delivered
> in time ({{too-late-packet-drop}}).

Reviewer 2: ```suggestion
In live mode, the sender is allowed to drop packets that cannot be delivered
in time ({{too-late-packet-drop}}).
```

Author: In case of live streaming, the sender is allowed to drop packets that cannot be delivered
in time ({{too-late-packet-drop}}).

Correct?

Reviewer 2: Yes, with one tiny comment. The expressions "in case of" and "in the case of" in English have a subtle difference when used. Not easy to explain, but "in **the** case of" usually sounds better. Not a big deal either way.

> To sum up, a combination of pacing control and live
> congestion control (LiveCC) modules controls SRT packets pacing based
> on the input rate and an overhead for packets retransmission in order
> to avoid congestion during fluctuations of the source bitrate.

Reviewer 2: ```suggestion
The combination of pacing control and live congestion control (LiveCC), based
on the input rate and an overhead for packets retransmission, helps avoid congestion 
during fluctuations of the source bitrate.
```

In addition to these mechanisms, in highly variable networks live mode
fairness can be accomplished outside SRT by controlling the bitrate
of the encoder at the input of the SRT sender. SRT provides a variaty
of network related statistics such as RTT estimation, the level of
packet loss, the number of packets dropped by SRT, etc., which can be

> used for making decisions and adjusting the bitrate in real time.

Reviewer 2: ```suggestion
In live mode over highly variable networks, fairness can be achieved by controlling 
the bitrate of the source encoder at the input of the SRT sender. SRT provides a 
variety of network related statistics, such as RTT estimation, packet loss levels, 
the number of packets dropped, etc., which can be used for making decisions 
and adjusting the bitrate in real time.
```

Author: **In case of live streaming** over highly variable networks, fairness can be achieved by controlling
the bitrate of the source encoder at the input of the SRT sender. SRT provides a
variety of network related statistics, such as RTT estimation, packet loss **level,**
the number of packets dropped, etc., which can be used for making decisions
and adjusting the bitrate in real time.

Is there a way to say "in case of live streaming" in other words?

Reviewer 2: Your changes are fine. Again, subjectively, there is the use of "in case...". You could also say "During live streaming . . ." or "When live streaming . . .".
Also, the word "estimation" usually refers to the act of performing a calculation where "estimate" is the result. This is a very subtle distinction, and using "estimation" as a noun is technically not wrong. More of a "common usage" thing.

### Configuring Maximum Bandwidth {#maxbw}

> There are several ways of configuring maximum bandwidth MAX_BW:
> 
> 1. By means of setting the value of maximum bandwidth MAX_BW
>    explicitly (MAXBW mode).
> 
>    The recommended default value is 1 Gbps. Note that the default value
>    is set only for live transmission mode.
> 
>    Note that this static setting is not well-suited to a variable input,
>    like when you change the bitrate on an encoder. Each time the input
>    bitrate is configured on the encoder, MAX_BW must also be reconfigured.
> 
>    In this mode, SRT does not follow the fluctuations in the encoder input
>    making it not suitable for variable bitrate (VBR) encoding use cases.
> 
> 2. By means of setting the SRT sender's input rate INPUT_BW and the
>    overhead OVERHEAD (INPUTBW mode).

Reviewer 2: ```suggestion
There are several ways of configuring maximum bandwidth (MAX_BW):

1. Set the value explicitly (MAXBW mode).

   The recommended default value is 1 Gbps. The default value
   is set only for live transmission mode.

   Note that this static setting is not well-suited to a variable input,
   like when you change the bitrate on an encoder. Each time the input
   bitrate is configured on the encoder, MAX_BW must also be reconfigured.

   In MAXBW mode, SRT does not follow any fluctuations in the encoder input
   making it unsuitable for variable bitrate (VBR) encoding.

2. Set the SRT sender's input rate (INPUT_BW) and OVERHEAD (INPUTBW mode).
```

   In this mode, SRT calculates the maximum bandwidth as follows:

   ~~~
   MAX_BW = INPUT_BW * (1 + OVERHEAD /100)
   ~~~

   Note that INPUTBW mode reduces to the MAXBW mode.

> 3. Measuring SRT sender's input rate internally and setting the
>    overhead OVERHEAD (ESTINPUTBW mode).

Reviewer 2: ```suggestion
3. ESTINPUTBW mode: Measure the SRT sender's input rate internally and set the
   OVERHEAD.
```

   In this mode, SRT adjusts the value of maximum bandwidth each time it
   gets the updated estimation of input rate EST_INPUT_BW:

   ~~~
   MAX_BW = EST_INPUT_BW * (1 + OVERHEAD /100)
   ~~~

> 4. A hybrid mode which takes the configured input rate INPUT_BW,
>    compares it against estimated value EST_INPUT_BW, and uses the
>    maximum of two (HYBRIDBW mode):

Reviewer 2: ```suggestion
4. HYBRIDBW mode -- Compare the configured input rate INPUT_BW 
   against the estimated bandwidth EST_INPUT_BW, and use the
   maximum of two:
```

   ~~~
   MAX_BW = max(INPUT_BW, EST_INPUT_BW) * (1 + OVERHEAD /100)
   ~~~

Note that the units of MAX_BW, INPUT_BW, and EST_INPUT_BW are bytes per
second. OVERHEAD is defined in %.

Both ESTINPUTBW and HYBRIDBW modes are recommended for setting the

> maximum bandwidth as it (MAX_BW) follows the fluctuations in

Reviewer 2: ```suggestion
maximum bandwidth (MAX_BW) as it follows the fluctuations in
```

SRT sender's input rate. However, there are certain considerations
that should be taken into account.

In INPUTBW mode, SRT takes as an input the rate that had been
configured as the expected output rate of an encoder (in terms
of bitrate for the packets including audio and overhead). But it is
normal for an encoder to occasionally overshoot. At low bitrate,
sometimes an encoder can be too optimistic and will output more bits
than expected. Under these conditions, SRT packets would not go out

> fast enough due to too low value of the configured bandwidth limitation.

Reviewer 2: ```suggestion
fast enough because the configured bandwidth limitation would be too low.
```

This is mitigated by calculating the bitrate internally (ESTINPUTBW
mode). SRT examines the packets being submitted and calculates the
input rate as a moving average. However, this introduces a bit of a
delay based on the content. It also means that if an encoder encounters
black screens or still frames, this would dramatically lower the
bitrate being measured, which would in turn reduce the SRT output rate.
And then, when the video picks up again, the input rate rises sharply.
SRT would not start up again fast enough on output because of the time
it takes to measure the speed. Packets might be accumulated in the

> SRT's sender buffer and, as a result, delayed causing their too late
> arrival at the decoder and possible drops by the receiver.

Reviewer 2: ```suggestion
SRT's sender buffer and delayed as a result, causing them to arrive too late 
at the decoder, and possible drops by the receiver.
```

> A proposed solution is to have SRT take the input rate configuration
> from the encoder, measure the actual input, and use whichever value
> is larger (HYBRIDBW mode). Combining the two methods overcomes
> the deficiencies of each.

Reviewer 2: ```suggestion
HYBRIDBW mode combines the two methods above and overcomes 
the deficiencies of each. SRT takes the input rate configuration from the 
encoder, measures the actual input, and uses whichever value is larger.
```

Reviewer: I also agree with this suggestion, as using the term "proposed solution" is too loud for such case. Each mode has its use case. Let us not put any of them in favor.

The following table shows the summary of modes, variables need to be
set (✓) or ignored (-) and formula for MAX_BW calculation for a

> particular mode:

Reviewer 2: ```suggestion
The following table shows a summary of the bandwidth configuration modes, 
the variables that need to be set (✓) or ignored (-), and the formula for 
calculating MAX-BW in each case:
```

~~~
| Mode / Variable | MAX_BW | INPUT_BW | OVERHEAD |
| --------------- | ------ | -------- | -------- |
| MAXBW           | ✓      | -        | -        |
| INPUTBW         | -      | ✓        | ✓        |
| ESTINPUTBW      | -      | -        | ✓        |
| HYBRIDBW        | -      | ✓        | ✓        |
~~~

### SRT's Default LiveCC Algorithm {#default-liveCC}

The main goal of the SRT's default LiveCC algorithm is to adjust the
minimum allowed packet sending period PKT_SND_PERIOD (and, as a result,
the maximum allowed sending rate) during transmission based on the

> average packet payload size AvgPayloadSize and maximum bandwidth
> MAX_BW.

Reviewer 2: ```suggestion
average packet payload size (AvgPayloadSize) and maximum bandwidth
(MAX_BW).
```

> On the sender side, there are three events LiveCC algorithm reacts to:
> on sending a data packet, on an acknowledgement (ACK) packet reception,
> and on a timeout event as described below.

Reviewer 2: ```suggestion
On the sender side, there are three events that the LiveCC algorithm reacts to:
(1) on sending a data packet, (2) on receiving an acknowledgement (ACK) packet, 
and (3) on a timeout event as described below.
```

> On sending a data packet (either original, or retransmitted), update
> the value of average packet payload size AvgPayloadSize:

Reviewer 2: ```suggestion
On sending a data packet (either original or retransmitted), update
the value of average packet payload size (AvgPayloadSize):
```

~~~
AvgPayloadSize = AvgPayloadSize * 0.875 + PacketPayloadSize * 0.125
~~~

where PacketPayloadSize is the payload size of a sent data packet, in bytes;

> the initial value of AvgPayloadSize is equal the maximum allowed packet
> payload size which can't be larger than 1456 bytes.

Reviewer 2: ```suggestion
the initial value of AvgPayloadSize is equal to the maximum allowed packet
payload size, which can't be larger than 1456 bytes.
```

> On an acknowledgement (ACK) packet reception:

Reviewer 2: ```suggestion
(2) On an acknowledgement (ACK) packet reception:
```

> Step 1. Calculate SRT packet size pktsize as the sum of average payload size
> AvgPayloadSize and SRT header ({{packet-structure}}), in bytes.

Reviewer 2: ```suggestion
Step 1. Calculate SRT packet size (pktsize) as the sum of average payload size
(AvgPayloadSize) and SRT header size ({{packet-structure}}), in bytes.
```

> Step 2. Calculate the minimum allowed packet sending period PKT_SND_PERIOD

Reviewer 2: ```suggestion
Step 2. Calculate the minimum allowed packet sending period (PKT_SND_PERIOD) 
```

as:

~~~
PKT_SND_PERIOD = pktsize * 1000000 / MAX_BW
~~~

where MAX_BW is the configured maximum bandwidth which limits the
bandwidth usage by SRT, in bytes per second; PKT_SND_PERIOD is
measured in microseconds.

> On a retransmission timeout (RTO) event, follow the same steps as
> described in "On an acknowledgement (ACK) packet reception" part.

Reviewer 2: ```suggestion
(3) On a retransmission timeout (RTO) event, follow the same steps as
described in method (1) above.
```

> RTO is the time that an acknowledgement is expected after a data packet
> is sent out. If there is no ACK after this amount of time elapsed, a

Reviewer 2: ```suggestion
RTO is the amount of time within which an acknowledgement is expected after a data packet
is sent out. If there is no ACK after this amount of time has elapsed, a
```

timeout event is triggered. Since SRT only acknowledges every SYN time
({{packet-acks}}), the value of retransmission timeout is
defined as follows:

~~~
RTO = RTT + 4 * RTTVar + 2 * SYN
~~~

where RTT is the round-trip time estimation, in microseconds, and
RTTVar is the variance of RTT estimation, in microseconds, reported

> by the receiver and smoothed at the sender side

Reviewer 2: smoothed?

Author: Currently, the EWMA is applied twice: 1. at the receiver side, then we send this value within an ACK packet; 2. at the sender side: we extract RTT from ACK and do EWMA again, that's why I writing here smoothed at the receiver side.

Author: See [this comment](https://github.com/Haivision/srt-rfc/pull/66#discussion_r492836186) also.

(see {{ctrl-pkt-ack}}, {{rtt}}).

Continuous timeout should increase the RTO value. In SRT, a counter

> (RexmitCount) is used to track the number of continuous timeout:

Reviewer 2: ```suggestion
(RexmitCount) is used to track the number of continuous timeouts:
```

~~~
RTO = RexmitCount * (RTT + 4 * RTTVar + 2 * SYN) + SYN
~~~

On the receiver side, on a loss report sending, the interval of sending
periodic NAK reports ({{packet-naks}}) is updated as follows:

~~~
NAKInterval = min((RTT + 4 * RTTVar) / 2, 20000)
~~~

where RTT and RTTVar are receiver's estimations (see {{ctrl-pkt-ack}},
{{rtt}}). The minimum value of NAKInterval is set to 20 milliseconds
in order to avoid sending periodic NAK reports too often under
low latency conditions.

## File Transfer Congestion Control (FileCC)

For file transfer ({{transmission-mode-msg}},
{{transmission-mode-buffer}}), any known file congestion control
algorithm like CUBIC {{RFC8312}} or BBR {{BBR}} can be applied,

> including SRT's default FileCC algorithm decsribed below

Reviewer 2: ```suggestion
including SRT's default FileCC algorithm described below
```

({{default-fileCC}}).

### SRT's Default FileCC Algorithm {#default-fileCC}

> SRT's default FileCC algorithm is a modified version of the UDT's

Reviewer 2: ```suggestion
SRT's default FileCC algorithm is a modified version of the UDT
```

native congestion control algorithm {{GuAnAO}}, {{GHG04b}}
designed for a bulk data transfer over networks with a large
bandwidth-delay product (BDP). It's a hybrid Additive Increase
Multiplicative Decrease (AIMD) algorithm, hence it adjusts both
congestion window size (CWND_SIZE) and packet sending period
(PKT_SND_PERIOD). The units of measurement for CWND_SIZE and
PKT_SND_PERIOD are packets and microseconds, respectively.

The algorithm controls sending rate by tuning packet sending period,
i.e. how often to send out the packets. The sending rate is increased
upon an acknowledgement (ACK) reception and decreased in case of
receiving a loss report (negative acknowledgement, or NAK).
Only full ACKs (not light) trigger the increase of sending rate.

> The congestion control includes two phases. It starts with a slow
> start phase where it probes the network to determine available
> bandwidth and the target sending rate for the desired operation
> mode: congestion avoidance. In this phase, if there is no congestion
> detected via loss reports, the sending rate is gradually increased.
> On the other hand, if a network congestion is detected, the algorithm
> slows down and decreases the sending rate to reduce following packet
> loss. The slow start phase runs exactly once at the beginning of a
> connection and stops when a packet loss occurs, congestion window
> size reaches its maximum value, or on a timeout event.

Reviewer 2: ```suggestion
SRT congestion control has two phases: "Slow Start" and "Congestion Avoidance". 
In the slow start phase the congestion control module probes the network to determine 
available bandwidth and the target sending rate for the next (operational) phase, which 
is congestion avoidance. In this phase, if there is no congestion detected via loss reports, 
the sending rate is gradually increased. Conversely, if a network congestion is detected, 
the algorithm decreases the sending rate to reduce subsequent packet loss. The slow start 
phase runs exactly once at the beginning of a connection, and stops when a packet loss 
occurs, when the congestion window size reaches its maximum value, or on a timeout event.
```

The detailed algorithm behaviour at both phases is described
in {{default-fileCC-slow-start}} and
{{default-fileCC-congestion-avoidance}}, respectively.

#### Slow Start {#default-fileCC-slow-start}

During the slow start phase, the packet sending period PKT_SND_PERIOD
is kept at 1 microsecond in order to send packets as fast as possible,

> but not with an infinite rate.

Reviewer 2: ```suggestion
but not at an infinite rate.
```

> The initial value of congestion window size CWND_SIZE is set to 16
> packets. CWND_SIZE has an upper threshold, the maximum allowed
> congestion window size MAX_CWND_SIZE, which is used for safety purposes
> (even if there is no packet loss, the slow start phase has to stop
> at a certain point). The threshold can be set to the maximum of
> the receiver buffer size (12 MB).

Reviewer 2: ```suggestion
The initial value of the congestion window size (CWND_SIZE) is set to 16
packets. CWND_SIZE has an upper threshold, which is the maximum allowed
congestion window size (MAX_CWND_SIZE), so that even if there is no packet 
loss, the slow start phase has to stop at a certain point. The threshold can be 
set to the maximum receiver buffer size (12 MB).
```

in time ({{too-late-packet-drop}}).

To sum up, a combination of pacing control and live
congestion control (LiveCC) modules controls SRT packets pacing based
on the input rate and an overhead for packets retransmission in order
to avoid congestion during fluctuations of the source bitrate.

In addition to these mechanisms, in highly variable networks live mode
fairness can be accomplished outside SRT by controlling the bitrate
of the encoder at the input of the SRT sender. SRT provides a variety
of network related statistics such as RTT estimation, the level of
packet loss, the number of packets dropped by SRT, etc., which can be
used for making decisions and adjusting the bitrate in real time.

### Configuring Maximum Bandwidth {#maxbw}

There are several ways of configuring maximum bandwidth MAX_BW:

1. By means of setting the value of maximum bandwidth MAX_BW
   explicitly (MAXBW mode).

   The recommended default value is 1 Gbps. Note that the default value
   is set only for live transmission mode.

   Note that this static setting is not well-suited to a variable input,
   like when you change the bitrate on an encoder. Each time the input

>    bitrate is configured on the encoder, MAX_BW must also be reconfigured.

Author: We currently have for MAXBW mode the following explanation:

_The recommended default value is 1 Gbps. The default value
   is set only for live streaming.

   Note that this static setting is not well-suited to a variable input,
   like when you change the bitrate on an encoder. Each time the input
   bitrate is configured on the encoder, MAX_BW must also be reconfigured.

   In MAXBW mode, SRT does not follow any fluctuations in the encoder input
   making it unsuitable for variable bitrate (VBR) encoding._

If we change the second paragraph to your suggestion, the last one seems to be incorrect too. The same idea as you described. Do you have a suggestion how to change the last paragraph or maybe we should simply remove it, because it's already written in the second paragraph?

>    bitrate is configured on the encoder, MAX_BW must also be reconfigured.

Reviewer: Simply removing the paragraph about "In MAXBW mode, SRT does not follow any fluctuations.." is a good idea.

   In this mode, SRT does not follow the fluctuations in the encoder input
   making it not suitable for variable bitrate (VBR) encoding use cases.

2. By means of setting the SRT sender's input rate INPUT_BW and the
   overhead OVERHEAD (INPUTBW mode).

   In this mode, SRT calculates the maximum bandwidth as follows:

   ~~~
   MAX_BW = INPUT_BW * (1 + OVERHEAD /100)
   ~~~

>    Note that INPUTBW mode reduces to the MAXBW mode.

Author: That's why I've written here: Note that INPUTBW mode reduces to the MAXBW mode.
Do you think I should explicitly write the same paragraph as for MAXBW mode?

>    Note that INPUTBW mode reduces to the MAXBW mode.

Reviewer: Maybe something like "Note that INPUTBW has the same behavior as the MAXBW mode, and the same restrictions apply".

The algorithm controls sending rate by tuning packet sending period,
i.e. how often to send out the packets. The sending rate is increased
upon an acknowledgement (ACK) reception and decreased in case of
receiving a loss report (negative acknowledgement, or NAK).
Only full ACKs (not light) trigger the increase of sending rate.

The congestion control includes two phases. It starts with a slow
start phase where it probes the network to determine available
bandwidth and the target sending rate for the desired operation
mode: congestion avoidance. In this phase, if there is no congestion
detected via loss reports, the sending rate is gradually increased.
On the other hand, if a network congestion is detected, the algorithm
slows down and decreases the sending rate to reduce following packet
loss. The slow start phase runs exactly once at the beginning of a
connection and stops when a packet loss occurs, congestion window
size reaches its maximum value, or on a timeout event.

The detailed algorithm behaviour at both phases is described
in {{default-fileCC-slow-start}} and
{{default-fileCC-congestion-avoidance}}, respectively.

> 

Reviewer 2: ```suggestion

As with LiveCC, SRT's default FileCC algorithm reacts to three events:
(1) sending a data packet, (2) receiving an acknowledgement (ACK) packet, 
and (3) a timeout event. These are described below as they apply to the 
congestion control phases.

```

#### Slow Start {#default-fileCC-slow-start}

During the slow start phase, the packet sending period PKT_SND_PERIOD
is kept at 1 microsecond in order to send packets as fast as possible,
but not with an infinite rate.
The initial value of congestion window size CWND_SIZE is set to 16
packets. CWND_SIZE has an upper threshold, the maximum allowed
congestion window size MAX_CWND_SIZE, which is used for safety purposes
(even if there is no packet loss, the slow start phase has to stop
at a certain point). The threshold can be set to the maximum of
the receiver buffer size (12 MB).

> On an acknowledgement (ACK) packet reception:

Reviewer 2: ```suggestion
(1) On an acknowledgement (ACK) packet reception:
```

> Step 1. If the time passed since the last time LastRCTime the sending rate
> was either increased, or kept, is less than RC_INTERVAL:

Reviewer 2: ```suggestion
Step 1. If the interval since the last time the sending rate was either increased or kept (LastRCTime) is less than RC_INTERVAL:
```

a. Keep the sending rate at the same level;

b. Stop.

    <CODE BEGINS>
    if (currTime - LastRCTime < RC_INTERVAL)
    {
        Keep the sending rate at the same level;
        Stop;
    }
    <CODE ENDS>

where currTime is the current time, in microseconds; LastRCTime is
the last time the sending rate was either increased, or kept,
in microseconds.

Step 2. Update the value of LastRCTime to the current time:

~~~
LastRCTime = currTime
~~~

Step 3. The size of congestion window CWND_SIZE is increased by the
difference in sequence numbers of the data packet being acknowledged
ACK_SEQNO and the last acknowledged data packet LAST_ACK_SEQNO:

~~~
CWND_SIZE += ACK_SEQNO - LAST_ACK_SEQNO
~~~

Step 4. The sequence number of the last acknowledged data packet
LAST_ACK_SEQNO is updated as follows:

~~~
LAST_ACK_SEQNO = ACK_SEQNO
~~~

> Step 5. If calculated at Step 3 congestion window size CWND_SIZE is greater
> than the upper threshold MAX_CWND_SIZE, slow start phase ends.

Reviewer 2: ```suggestion
Step 5. If the congestion window size CWND_SIZE calculated at Step 3 is greater
than the upper threshold MAX_CWND_SIZE, slow start phase ends.
```

Set the packet sending period PKT_SND_PERIOD as follows:

    <CODE BEGINS>
    if (RECEIVING_RATE > 0)
        PKT_SND_PERIOD = 1000000 / RECEIVING_RATE;
    else
        PKT_SND_PERIOD = CWND_SIZE / (RTT + RC_INTERVAL);
    <CODE ENDS>

where

- RECEIVING_RATE is the rate at which packets are being received,

> in packets per second, reported by the receiver and smoothed at

Reviewer 2: smoothed?

Author: the same explanation as for RTT

the sender side (see {{ctrl-pkt-ack}}, {{bandwidth-estimation}});
- RTT is the round-trip time estimation, in microseconds,

> reported by the receiver and smoothed at the sender side (see

Reviewer 2: smoothed?

Author: the same explanation as for the very first smoothed RTT

{{ctrl-pkt-ack}}, {{rtt}});
- RC_INTERVAL is the fixed rate control interval, in microseconds.
RC_INTERVAL of SRT is SYN, or synchronization time interval, which
is 0.01 second. An ACK in SRT is sent every fixed time interval.
The maximum and default ACK time interval is SYN.
See {{packet-acks}} for details.

> On a loss report (NAK) packet reception:

Reviewer 2: ```suggestion
(2) On a loss report (NAK) packet reception:
```

- Slow start phase ends;

> - Set the packet sending period PKT_SND_PERIOD as described in Step 5,
> "On an acknowledgement (ACK) packet reception" part.

Reviewer 2: ```suggestion
- Set the packet sending period PKT_SND_PERIOD as described in Step 5 of section (1) above.
```

> On a retransmission timeout (RTO) event:

Reviewer 2: ```suggestion
(3) On a retransmission timeout (RTO) event:
```

- Slow start phase ends;

> - Set the packet sending period PKT_SND_PERIOD as described in Step 5,
> "On an acknowledgement (ACK) packet reception" part.

Reviewer 2: ```suggestion
- Set the packet sending period PKT_SND_PERIOD as described in Step 5 of section (1) above.
```

#### Congestion Avoidance {#default-fileCC-congestion-avoidance}

Once the slow start phase ends, the algorithm enters the congestion
avoidance phase and behaves as described below.

> On an acknowledgement (ACK) packet reception:

Reviewer 2: ```suggestion
(1) On an acknowledgement (ACK) packet reception:
```

> Step 1. If the time passed since the last time LastRCTime the sending rate
> was either increased, or kept, is less than RC_INTERVAL:

Reviewer 2: ```suggestion
Step 1. If the interval since the last time the sending rate was either increased or kept (LastRCTime) is less than RC_INTERVAL:
```

b. Set the value of bLoss to False;

c. Stop.

bLoss flag is equal to True if a packet loss has happened since the
last sending rate increase. Initial value: False.

Step 5. If there is no packet loss reported by the receiver (bLoss=False),
calculate PKT_SND_PERIOD as follows:

    <CODE BEGINS>
    inc = 0;

    lossBandwidth = 2 * (1000000 / LastDecPeriod);
    linkCapacity = min(lossBandwidth, EST_LINK_CAPACITY);
    B = linkCapacity - 1000000 / PKT_SND_PERIOD;

    if ((PKT_SND_PERIOD > LastDecPeriod) && ((linkCapacity / 9) < B))
        B = linkCapacity / 9;
    if (B <= 0)
        inc = 1 / S;
    else
    {
        inc = pow(10.0, ceil(log10(B * S * 8))) * 0.0000015 / S;
        inc = max(inc, 1 / S);
    }

    PKT_SND_PERIOD = (PKT_SND_PERIOD * RC_INTERVAL) / 
                      (PKT_SND_PERIOD * inc + RC_INTERVAL);
    <CODE ENDS>

where

- LastDecPeriod is the value of PKT_SND_PERIOD right before the last
sending rate decrease has happened (on a loss report (NAK) packet
reception), in microseconds. The initial value of LastDecPeriod
is set to 1 microsecond;
- EST_LINK_CAPACITY is the estimated link capacity reported

> by the receiver within an ACK packet and smoothed at the sender side

Reviewer 2: smoothed?

Author: still the same explanation )

({{bandwidth-estimation}), in packets per second;
- B is the estimated available bandwidth, in packets per second;
- S is the SRT packet size (in terms of IP payload) in bytes.
SRT treats 1500 bytes as a standard packet size.

> The detailed explanation of the formulas used to calculate the increase
> in sending rate inc can be found in {{GuAnAO}}. The UDT's available
> bandwidth B estimation has been modified to take into account the
> bandwidth registered at the moment of packet loss since the estimated
> link capacity EST_LINK_CAPACITY reported by the receiver may
> overestimate the real one significantly.

Reviewer 2: ```suggestion
A detailed explanation of the formulas used to calculate the increase
in sending rate can be found in {{GuAnAO}}. UDT's available
bandwidth estimation has been modified to take into account the
bandwidth registered at the moment of packet loss, since the estimated
link capacity EST_LINK_CAPACITY reported by the receiver may
overestimate the actual link capacity significantly.
```

Step 6. If the value of maximum bandwidth MAX_BW defined in
{{liveCC}} is set, limit the value of PKT_SND_PERIOD to the
minimum allowed period, if necessary:

    <CODE BEGINS>
    if (MAX_BW)
        MIN_PERIOD = 1000000 / (MAX_BW / S);

        if (PKT_SND_PERIOD < MIN_PERIOD)
            PKT_SND_PERIOD = MIN_PERIOD;
    <CODE ENDS> 

> Note that in file transmission mode there is an opportunity to define
> the maximum allowed bandwidth MAX_BW for SRT which limits the minimum
> possible packets' inter-sending period. Only the usage of MAXBW mode
> is possible (see {{maxbw}}). In contrast with live transmission
> mode, there is no default value set for MAX_BW and transmission rate
> is not limited if not set explicitly.

Reviewer 2: ```suggestion
Note that in file transmission mode the the maximum allowed bandwidth (MAX_BW) 
for SRT can be defined. This  limits the minimum possible interval between packets 
sent. Only the usage of MAXBW mode is possible (see {{maxbw}}). In contrast with 
live transmission mode, there is no default value set for MAX_BW, and the transmission 
rate is not limited if not set explicitly.
```

Author: Note that in file transmission mode the maximum allowed bandwidth (MAX_BW)
for SRT can be defined. This limits the minimum possible **packets' inter-sending period**.
Only the usage of MAXBW mode is possible ({{maxbw}}). In contrast with
live transmission mode, there is no default value set for MAX_BW, and the transmission
rate is not limited if not set explicitly.

I need to say "inter-sending period" here.

Reviewer 2: Okay. Just so that you know "inter-sending" is not really a word!  ;-)

Also, in your comment above, you said "...we have two modes in SRT: message and buffer, ...". So maybe change "Note that in file transmission mode..." to "During file transmission..."?

Author: Okay, interesting. I know that there is a word "inter-arrival" ))) Will use you original suggestion + correct live/file modes.

> On a loss report (NAK) packet reception:

Reviewer 2: ```suggestion
(2) On a loss report (NAK) packet reception:
```

Step 1. Set the value of flag bLoss equal to True.

Step 2. If the current loss ratio estimated by the sender is less than 2%:

a. Keep the sending rate at the same level;

> b. Update the value of LastDecPeriod: LastDecPeriod = PKT_SND_PERIOD;

Reviewer 2: ```suggestion
b. Update the value of LastDecPeriod: 

~~~
LastDecPeriod = PKT_SND_PERIOD;
~~~

```

period PKT_SND_PERIOD;

b. Increase the value of packet sending period:

~~~
PKT_SND_PERIOD = PKT_SND_PERIOD * 1.03
~~~

c. Update AvgNAKNum:

~~~
AvgNAKNum = AvgNAKNum * 0.97 + NAKCount * 0.03
~~~

d. Reset NAKCount and DecCount values to 1;

e. Record the current largest sent sequence number LastDecSeq;

f. Compute DecRandom to a random (uniform distribution) number
between 1 and AvgNAKNum. If DecRandom < 1: DecRandom = 1;

g. Stop;

where

- AvgNAKNum is the average number of NAKs during a congestion period.
  Initial value: 0;
- NAKCount is the number of NAKs received so far in the current
  congestion period. Initial value: 0;
- DecCount means the number of times that the sending rate has been
  decreased during the congestion period. Initial value: 0;
- DecRandom is a random number used to decide if the rate should be
  decreased or not for the following NAKs (not the first one) during
  the congestion period. DecRandom is a random number between 1 and
  the average number of NAKs per congestion period (AvgNAKNum).

Congestion period is defined as the time between two NAKs in which
the biggest lost packet sequence number carried in the NAK is greater
than the LastDecSeq.

> A slight modification of coefficients used in the above mentioned
> formulas has been made in order to reduce the value of sending
> rate decrease.

Reviewer 2: ```suggestion
The coefficients used in the formulas above have been slightly modified 
to reduce the amount by which the sending rate decreases.
```

Step 4. If DecCount <= 5, and NAKCount == DecCount * DecRandom:

a. Update SND period: SND = SND * 1.03;

b. Increase DecCount and NAKCount by 1;

c. Record the current largest sent sequence number (LastDecSeq).

#### Link Capacity and Receiving Rate Estimation {#bandwidth-estimation}

> Link capacity and receiving rate estimations, in packets/bytes per second,
> are calculated at the receiver side during the file transmission.
> It's worth noting that receiving rate estimation is used only during
> the slow start period of the file congestion control algorithm
> ({{default-fileCC-slow-start}}), however it's available during
> the whole data transmission period.

Reviewer 2: ```suggestion
Estimates of link capacity and receiving rate, in packets/bytes per second,
are calculated at the receiver side during file transmission. It's worth noting 
that the receiving rate estimate, while available during the entire data transmission 
period, is used only during the slow start phase of the file congestion control 
algorithm ({{default-fileCC-slow-start}}).
```

> The latest estimation obtained before the end of the slow start

Reviewer 2: ```suggestion
The latest estimate obtained before the end of the slow start
```

period is used by the sender as a reference maximum speed
to continue data transmission without further congestion.
Link capacity is estimated all the time and used primarily
(as well as packet loss ratio and other protocol statistics)

> for smoothed sending rate adjustments during the transmission process.

Reviewer 2: smoothed?

Author: maybe we should simply delete "smoothed" from here - used primarily for sending rate adjustments ...
I meant "smooth" here if it helps.

Author: I see this below:
_for making sending rate adjustments  during the transmission process 
by applying an exponential moving average (EMA) to smooth the data._

This isn't right for this case, but this part
_by applying an exponential moving average (EMA) to smooth the data at the sender side_
is correct for all the other cases where I apply smoothed term

Plus better to use
_exponentially weighted moving average (EWMA)_

Reviewer 2: If we clarify that "smoothing" means "applying an exponentially weighted moving average (EWMA)" early in the Congestion Control section, then it is perfectly okay to use just "smoothing" later on.

> The receiver records the inter-arrival time of each packet
> (time delta with the previous data packet) which is further
> used in the models to estimate bandwidth and receiving speed.
> The communication between receiver and sender happens by means
> of acknowledgment packets which are sent regularly
> (each 10 milliseconds) and contain some control information
> as well as bandwidth and delivery rate estimations.

Reviewer 2: ```suggestion
As each data packet arrives, the receiver records the time delta with respect to 
the arrival of the previous data packet, which is used  to estimate bandwidth and 
receiving speed (delivery rate). This and other control information is communicated 
to the sender by means of acknowledgment (ACK) packets sent every 10 milliseconds.
```

> At the sender side, upon receiving a new value, a smoothed average

Reviewer 2: smoothed?

Author: the same explanation as for RTT, receiving rate and link capacity reported by receiver via ACK packets
upon ACK arrival, at the sender side we apply EWMA one more time to all these estimates

> is used to update the latest estimation maintained at the sender side.

Reviewer 2: ```suggestion
is used to update the latest estimate maintained at the sender side.
```

> It is important to note that for bandwidth estimation only data
> probing packets are taken into account while all the data packets
> (both data and data probing) are used for receiving speed estimation.
> The idea behind packet pairs technique is to send the groups
> of back-to-back packets, i.e., probing packet pairs, to a server
> thus making it possible to measure the minimum interval in receiving
> the consecutive packets.

Reviewer 2: ```suggestion
It is important to note that for bandwidth estimation only data
probing packets are taken into account, while all data packets
(both data and data probing) are used for estimating receiving 
speed. Data probing refers to the use of the packet pairs technique, 
whereby pairs of probing packets are sent to a server back-to-back, 
thus making it possible to measure the minimum interval in receiving 
consecutive packets.
```

Author: What do you mean by "back-to-back" here?

Reviewer 2: No delay between sending packets in a probing pair.
![SRT_PacketProbe_2018-09-10](https://user-images.githubusercontent.com/28537807/94163046-634eb800-fe55-11ea-8ecc-25083bbbaad2.png)

## SRT Packet Pacing and Live Congestion Control (LiveCC) {#liveCC}

> To ensure smooth video playback on a receiving peer, SRT in case of
> live streaming must control the sender's buffer level to

Reviewer 2: ```suggestion
To ensure smooth video playback on a receiving peer during  
live streaming, SRT must control the sender's buffer level to
```

to continue data transmission without further congestion.
Link capacity is estimated all the time and used primarily
(as well as packet loss ratio and other protocol statistics)

> for smoothed sending rate adjustments during the transmission process.

Reviewer 2: ```suggestion
for making sending rate adjustments by applying a smooth moving average 
(smoothing) during the transmission process.
```
